import { Component, OnInit } from '@angular/core';

@Component({
  selector: 'app-software',
  templateUrl: './software.component.html',
  styleUrls: ['./software.component.scss']
})
export class SoftwareComponent implements OnInit {
  software: Object[] = [{ "name": "CViSB Data Portal", "description": "The CViSB Data Portal provides access to all datasets and patient metadata generated or curated by the Center for Viral Systems Biology, which studies Ebola, Lassa Fever, and COVID-19.", "creator": "CViSB", "url": "https:\/\/data.cvisb.org\/home", "Funder name": "NIAID", "Identifier": "U19-AI135995", "applicationCategory": ["Database portal", "Web application"], "applicationSubCategory": null, "additonalType": null, "citation": null, "Citation URL": null, "license": null, "Species": null, "Infectious Agent": "EBOLA VIRUS and LASSA FEVER VIRUS", "Infectious Disease": null, "dateCreated": null, "dateModified": null, "Date published": null, "softwareVersion": null, "Measurement Technique": null, "Programming Language": null, "Unnamed: 22": null, "Parent organization (remove)": "CViSB", "Grant description (remove)": null, "Funder name (should all be niaid)": "Andersen, Kristian Graugaard", "Grant URL": "https:\/\/reporter.nih.gov\/project-details\/9455502", "funder": ["NIAID"], "programmingLanguage": ["JavaScript"] }, { "name": "Floki", "description": "Floki is a tool under active development that can be used to visualize and interpret metagenomic data using the hierarchical taxonomy structure. It can be used to identify and filter out contaminants. It can also be used to visualize batch effects by grouping multiple samples together. Annotations from other sources at every taxon level can be combined with sequencing results to narrow down the search space for taxons of interest.", "creator": "CViSB", "url": "https:\/\/github.com\/cvisb\/floki", "Funder name": "NIAID", "Identifier": "U19-AI135995", "applicationCategory": ["Command-line tool", "Web application"], "applicationSubCategory": ["Genomic Sequencing"], "additonalType": null, "citation": "None", "Citation URL": "None", "license": "None", "Species": "ALL PATHOGENIC SPECIES", "Infectious Agent": "ALL PATHOGENIC SPECIES", "Infectious Disease": "ALL INFECTIOUS DISEASES", "dateCreated": "Feb 9, 2017", "dateModified": "Jun 26, 2019", "Date published": "Feb 9, 2017", "softwareVersion": "0", "Measurement Technique": null, "Programming Language": "C++;JavaScript", "Unnamed: 22": null, "Parent organization (remove)": "CViSB", "Grant description (remove)": null, "Funder name (should all be niaid)": "Andersen, Kristian Graugaard", "Grant URL": "https:\/\/reporter.nih.gov\/project-details\/9455502", "funder": ["NIAID"], "programmingLanguage": ["C++", "JavaScript"] }, { "name": "iVar", "description": "iVar is a computational package that contains functions broadly useful for viral amplicon-based sequencing. Additional tools for metagenomic sequencing are actively being incorporated into iVar. While each of these functions can be accomplished using existing tools, iVar contains an intersection of functionality from multiple tools that are required to call iSNVs and consensus sequences from viral sequencing data across multiple replicates. We implemented the following functions in iVar: (1) trimming of primers and low-quality bases, (2) consensus calling, (3) variant calling - both iSNVs and insertions\/deletions, and (4) identifying mismatches to primer sequences and excluding the corresponding reads from alignment files. An amplicon-based sequencing framework for accurately measuring intrahost virus diversity using PrimalSeq and iVar, Genome Biology 2019 20:8 Nathan D Grubaugh, Karthik Gangavarapu, Joshua Quick, Nathaniel L Matteson, Jaqueline Goes De Jesus, Bradley J Main, Amanda L Tan, Lauren M Paul, Doug E Brackney, Saran Grewal, Nikos Gurfield, Koen KA Van Rompay, Sharon Isern, Scott F Michael, Lark L Coffey, Nicholas J Loman, Kristian G Andersen bioRxiv doi: https:\/\/doi.org\/10.1101\/383513", "creator": "CViSB", "url": "https:\/\/github.com\/andersen-lab\/ivar", "Funder name": "NIAID", "Identifier": "U19-AI135995", "applicationCategory": ["Command-line tool"], "applicationSubCategory": ["Genomic Sequencing"], "additonalType": null, "citation": "Grubaugh, N.D., Gangavarapu, K. et al. An amplicon-based sequencing framework for accurately measuring intrahost virus diversity using PrimalSeq and iVar. Genome Biol 20, 8 (2019).\n", "Citation URL": "https:\/\/doi.org\/10.1186\/s13059-018-1618-7", "license": "GNU GPL v3", "Species": null, "Infectious Agent": "VIRUSES", "Infectious Disease": "VIRAL DISEASES", "dateCreated": "Dec 20, 2018", "dateModified": "Feb 5, 2021", "Date published": "Dec 20, 2018", "softwareVersion": "1.3.1", "Measurement Technique": null, "Programming Language": "C++", "Unnamed: 22": null, "Parent organization (remove)": "CViSB", "Grant description (remove)": null, "Funder name (should all be niaid)": "Andersen, Kristian Graugaard", "Grant URL": "https:\/\/reporter.nih.gov\/project-details\/9455502", "funder": ["NIAID"], "programmingLanguage": ["C++"] }, { "name": "outbreak.info", "description": "In response to the current outbreak of SARS-CoV-2 (the virus that causes COVID-19), researchers worldwide have been generating and openly sharing data, publications, reagents, code, protocols, and more. Broad sharing of these resources improves the speed and efficiency of science. Unfortunately, there are no uniform standards and repositories for collecting all this information in one place. Outbreak.info aggregates data across scientific sources, providing tools to meet three major aims: (1) Track trends in COVID-19 cases and deaths, (2) Track daily developments in SARS-CoV-2 variants, and (3) Integrate publications, preprints, clinical trials, datasets, protocols, and other resources into one searchable library of COVID-19 research", "creator": "CViSB", "url": "https:\/\/outbreak.info\/", "Funder name": "NIAID", "Identifier": "U19-AI135995", "applicationCategory": ["Web application", "Bioinformatics portal", "Database portal"], "applicationSubCategory": null, "additonalType": null, "citation": null, "Citation URL": null, "license": "GNU GPL v3", "Species": null, "Infectious Agent": "SARS-COV-2", "Infectious Disease": "COVID-19", "dateCreated": "Feb 25, 2021", "dateModified": "Jul 27, 2021", "Date published": null, "softwareVersion": null, "Measurement Technique": null, "Programming Language": "JavaScript", "Unnamed: 22": null, "Parent organization (remove)": "CViSB", "Grant description (remove)": null, "Funder name (should all be niaid)": "Andersen, Kristian Graugaard", "Grant URL": "https:\/\/reporter.nih.gov\/project-details\/9455502", "funder": ["NIAID"], "programmingLanguage": ["JavaScript"] }, { "name": "api.outbreak.info", "description": "An application programming interface (API) that enables programmatic access to data available on outbreak.info", "creator": "CViSB", "url": "https:\/\/github.com\/outbreak-info\/outbreak.api", "Funder name": "NIAID", "Identifier": "U19-AI135995", "applicationCategory": ["Web API"], "applicationSubCategory": null, "additonalType": null, "citation": null, "Citation URL": null, "license": "GNU GPL v3", "Species": null, "Infectious Agent": "SARS-COV-2", "Infectious Disease": "COVID-19", "dateCreated": "Mar 19, 2020", "dateModified": "Jul 6, 2021", "Date published": "Mar 19, 2020", "softwareVersion": null, "Measurement Technique": null, "Programming Language": "Python", "Unnamed: 22": null, "Parent organization (remove)": "CViSB", "Grant description (remove)": null, "Funder name (should all be niaid)": "Andersen, Kristian Graugaard", "Grant URL": "https:\/\/reporter.nih.gov\/project-details\/9455502", "funder": ["NIAID"], "programmingLanguage": ["Python"] }, { "name": "Bjorn", "description": "A suite of miscellaneous tools that can be used to: 1) generate information for large-scale genomic surveillance of SARS-CoV-2 sequences. This functionality relies on external tools such as datafunk, minimap2, and pandas. 2) prepare results and data files from SARS-CoV-2 sequencing analysis for release to public databases such as GISAID, Google Cloud, and GitHub", "creator": "CViSB", "url": "https:\/\/github.com\/andersen-lab\/bjorn", "Funder name": "NIAID", "Identifier": "U19-AI135995", "applicationCategory": ["Command-line tool"], "applicationSubCategory": ["Genomic Sequencing"], "additonalType": null, "citation": null, "Citation URL": null, "license": "GNU GPL v3", "Species": null, "Infectious Agent": null, "Infectious Disease": null, "dateCreated": null, "dateModified": "Jul 27, 2021", "Date published": "January 22, 2021", "softwareVersion": null, "Measurement Technique": null, "Programming Language": null, "Unnamed: 22": null, "Parent organization (remove)": null, "Grant description (remove)": null, "Funder name (should all be niaid)": null, "Grant URL": null, "funder": ["NIAID"], "programmingLanguage": null }, { "name": "R package for outbreak.info", "description": "An R package that acts as a convenient client to the outbreak.info API. ", "creator": "CViSB", "url": "https:\/\/github.com\/outbreak-info\/R-outbreak-info", "Funder name": "NIAID", "Identifier": "U19-AI135995", "applicationCategory": ["Command-line tool"], "applicationSubCategory": null, "additonalType": null, "citation": null, "Citation URL": null, "license": "GNU GPL v3", "Species": null, "Infectious Agent": "SARS-COV-2", "Infectious Disease": "COVID-19", "dateCreated": "May 10, 2020", "dateModified": "Jun 28, 2021", "Date published": "May 10, 2020", "softwareVersion": null, "Measurement Technique": null, "Programming Language": "R", "Unnamed: 22": null, "Parent organization (remove)": "CViSB", "Grant description (remove)": null, "Funder name (should all be niaid)": "Andersen, Kristian Graugaard", "Grant URL": "https:\/\/reporter.nih.gov\/project-details\/9455502", "funder": ["NIAID"], "programmingLanguage": ["R"] }, { "name": "BEAST", "description": "BEAST is a cross-platform program for Bayesian analysis of molecular sequences using MCMC. It is entirely orientated towards rooted, time-measured phylogenies inferred using strict or relaxed molecular clock models. It can be used as a method of reconstructing phylogenies but is also a framework for testing evolutionary hypotheses without conditioning on a single tree topology. BEAST uses MCMC to average over tree space, so that each tree is weighted proportional to its posterior probability. We include a simple to use user-interface program for setting up standard analyses and a suit of programs for analysing the results.", "creator": "CViSB", "url": "https:\/\/github.com\/beast-dev\/beast-mcmc", "Funder name": "European Union Seventh Framework Programme for research, technological development and demonstration;European Union Seventh Framework Programme for research, technological development and demonstration;Wellcome Trust;NSF;NIH;NIH;NIAID;", "Identifier": "278433-PREDEMICS;725422-ReservoirDOCS;206298\/Z\/17\/Z;1264153;R01 HG006139; R01 AI107034;U19 AI135995.", "applicationCategory": ["Command-line tool"], "applicationSubCategory": null, "additonalType": null, "citation": null, "Citation URL": "https:\/\/doi.org\/10.1093\/ve\/vey016", "license": "LGPL v3", "Species": null, "Infectious Agent": "ALL", "Infectious Disease": "ALL", "dateCreated": null, "dateModified": null, "Date published": null, "softwareVersion": null, "Measurement Technique": null, "Programming Language": "Java", "Unnamed: 22": null, "Parent organization (remove)": null, "Grant description (remove)": null, "Funder name (should all be niaid)": null, "Grant URL": null, "funder": ["European Union Seventh Framework Programme for research, technological development and demonstration", "European Union Seventh Framework Programme for research, technological development and demonstration", "Wellcome Trust", "NSF", "NIH", "NIH", "NIAID", ""], "programmingLanguage": ["Java"] }, { "name": "RBeast", "description": "RBeast is an R package for working with output files from BEAST and BEAST2. This package consists of four main functions: (1) Use beautier to create BEAST2 input (.xml) files. (2) Use beastier to run BEAST2. (3) Use tracerer to parse BEAST2 output (.log, .trees, etc) files. (4) Use BEASTmasteR for tip-dating analyses using fossils as dated terminal taxa.", "creator": "CViSB", "url": "https:\/\/github.com\/beast-dev\/RBeast", "Funder name": "NSF;NIAID;", "Identifier": "1264153;U19-AI135995", "applicationCategory": ["Command-line tool"], "applicationSubCategory": null, "additonalType": null, "citation": null, "Citation URL": null, "license": "MIT", "Species": null, "Infectious Agent": "ALL", "Infectious Disease": "ALL", "dateCreated": null, "dateModified": null, "Date published": null, "softwareVersion": null, "Measurement Technique": null, "Programming Language": "R", "Unnamed: 22": null, "Parent organization (remove)": null, "Grant description (remove)": null, "Funder name (should all be niaid)": null, "Grant URL": null, "funder": ["NSF", "NIAID", ""], "programmingLanguage": ["R"] }, { "name": "Beagle", "description": "BEAGLE is a high-performance library that can perform the core calculations at the heart of most Bayesian and Maximum Likelihood phylogenetics packages. It can make use of highly-parallel processors such as those in graphics cards (GPUs) found in many PCs. The project involves an open API and fast implementations of a library for evaluating phylogenetic likelihoods (continuous time Markov processes) of biomolecular sequence evolution. The aim is to provide high performance evaluation \u2018services\u2019 to a wide range of phylogenetic software, both Bayesian samplers and Maximum Likelihood optimizers. This allows these packages to make use of implementations that make use of optimized hardware such as graphics processing units.", "creator": "CViSB", "url": "https:\/\/github.com\/beagle-dev\/beagle-lib", "Funder name": "NFS;NFS;NFS;NFS;NFS;NIH;NIH;NIAID;Wellcome Trust;European Research Council;", "Identifier": "IIS-1251151;DMS-1264153;DBI-1356562;DBI-1661443;DEB-1354146;R01-HG006139;R01-AI107034;U19-AI135995; 206298\/Z\/17\/Z;725422-ReservoirDOCS", "applicationCategory": ["Library"], "applicationSubCategory": null, "additonalType": null, "citation": null, "Citation URL": "https:\/\/doi.org\/10.1093\/sysbio\/syr100", "license": "LGPL v3", "Species": null, "Infectious Agent": "ALL", "Infectious Disease": "ALL", "dateCreated": null, "dateModified": null, "Date published": null, "softwareVersion": null, "Measurement Technique": null, "Programming Language": "C++", "Unnamed: 22": null, "Parent organization (remove)": null, "Grant description (remove)": null, "Funder name (should all be niaid)": null, "Grant URL": null, "funder": ["NFS", "NFS", "NFS", "NFS", "NFS", "NIH", "NIH", "NIAID", "Wellcome Trust", "European Research Council", ""], "programmingLanguage": ["C++"] }, { "name": "Tracer", "description": "Tracer is a software package for visualising and analysing the MCMC trace files generated through Bayesian phylogenetic inference. Tracer provides kernel density estimation, multivariate visualisation, demographic trajectory reconstruction, conditional posterior distribution summary and more. Tracer can read output files from MrBayes, BEAST, BEAST2, RevBayes, Migrate, LAMARC and and possibly other MCMC programs from other domains.", "creator": "CViSB", "url": "https:\/\/github.com\/beast-dev\/tracer", "Funder name": "NIAID", "Identifier": "U19-AI135995", "applicationCategory": ["Desktop application"], "applicationSubCategory": null, "additonalType": null, "citation": "Rambaut A, Drummond AJ, Xie D, Baele G and Suchard MA (2018) Posterior summarisation in Bayesian phylogenetics using Tracer 1.7. Systematic Biology.", "Citation URL": "https:\/\/doi.org\/10.1093\/sysbio\/syy032", "license": null, "Species": null, "Infectious Agent": "ALL", "Infectious Disease": "ALL", "dateCreated": null, "dateModified": null, "Date published": null, "softwareVersion": null, "Measurement Technique": null, "Programming Language": "Java", "Unnamed: 22": null, "Parent organization (remove)": null, "Grant description (remove)": null, "Funder name (should all be niaid)": null, "Grant URL": null, "funder": ["NIAID"], "programmingLanguage": ["Java"] }, { "name": "Clonify", "description": "Clonify is a software package that is able to perform unseeded lineage assignment on very large sets of antibody sequences. Defining the dynamics and maturation processes of antibody clonal lineages is crucial to understanding the humoral response to infection and immunization. Comprehensive study of antibody lineages has been limited by the lack of an accurate clonal lineage assignment algorithm capable of operating on next-generation sequencing datasets.", "creator": "CViSB", "url": "https:\/\/github.com\/briney\/clonify-python", "Funder name": "NIAID", "Identifier": "U19-AI135995", "applicationCategory": null, "applicationSubCategory": null, "additonalType": null, "citation": "Briney, B., Le, K., Zhu, J. et al. Clonify: unseeded antibody lineage assignment from next-generation sequencing data. Sci Rep 6, 23901 (2016). https:\/\/doi.org\/10.1038\/srep23901", "Citation URL": "https:\/\/www.nature.com\/articles\/srep23901", "license": null, "Species": null, "Infectious Agent": null, "Infectious Disease": null, "dateCreated": null, "dateModified": null, "Date published": null, "softwareVersion": null, "Measurement Technique": null, "Programming Language": "Python", "Unnamed: 22": null, "Parent organization (remove)": null, "Grant description (remove)": null, "Funder name (should all be niaid)": null, "Grant URL": null, "funder": ["NIAID"], "programmingLanguage": ["Python"] }, { "name": "AbStar", "description": "AbStar is a software package that is able to perform VDJ assignments and primary annotation of antibody and TCR sequencing data. Scalable from a single sequence to billions of sequences and conforms with AIRR data formatting standards.", "creator": "CViSB", "url": "https:\/\/github.com\/briney\/abstar", "Funder name": "NIAID", "Identifier": "U19-AI135995", "applicationCategory": null, "applicationSubCategory": null, "additonalType": null, "citation": null, "Citation URL": null, "license": null, "Species": null, "Infectious Agent": null, "Infectious Disease": null, "dateCreated": null, "dateModified": null, "Date published": null, "softwareVersion": null, "Measurement Technique": null, "Programming Language": "Python", "Unnamed: 22": null, "Parent organization (remove)": null, "Grant description (remove)": null, "Funder name (should all be niaid)": null, "Grant URL": null, "funder": ["NIAID"], "programmingLanguage": ["Python"] }, { "name": "AbUtils", "description": "AbUtils is a software package that provides programming primitives and utilities for interactive sequence processing, analysis, and visualization of BCR and TCR sequencing data. Conforms with AIRR data formatting standards.", "creator": "CViSB", "url": "https:\/\/github.com\/briney\/abutils", "Funder name": "NIAID", "Identifier": "U19-AI135995", "applicationCategory": null, "applicationSubCategory": null, "additonalType": null, "citation": null, "Citation URL": null, "license": null, "Species": null, "Infectious Agent": null, "Infectious Disease": null, "dateCreated": null, "dateModified": null, "Date published": null, "softwareVersion": null, "Measurement Technique": null, "Programming Language": "Python", "Unnamed: 22": null, "Parent organization (remove)": null, "Grant description (remove)": null, "Funder name (should all be niaid)": null, "Grant URL": null, "funder": ["NIAID"], "programmingLanguage": ["Python"] }, { "name": "Single Cell Analysis of B cells (SCAB)", "description": "This package of tools, named Single Cell Analysis of B cells (scab), is part of the broader ab[x] toolkit for antibody sequence and repertoire analysis and comprises three primary functions: 1) integration of single cell RNA-seq, AgBCs and natively-paired adaptive immune receptor sequence data; 2) provide a means for identifying antigen-positive cells from AgBC data using a probabilistic classifier; and 3) exploration and visualization of data-rich single cell profiles", "creator": "CViSB", "url": "http://github.com/briney/scab", "Funder name": "NIAID", "Identifier": "U19-AI135995", "applicationCategory": null, "applicationSubCategory": null, "additonalType": null, "citation": null, "Citation URL": null, "license": null, "Species": null, "Infectious Agent": null, "Infectious Disease": null, "dateCreated": null, "dateModified": null, "Date published": null, "softwareVersion": null, "Measurement Technique": null, "Programming Language": "Python", "Unnamed: 22": null, "Parent organization (remove)": null, "Grant description (remove)": null, "Funder name (should all be niaid)": null, "Grant URL": null, "funder": ["NIAID"], "programmingLanguage": ["Python"] }];

  constructor() { }

  ngOnInit(): void {
  }

}
