<div class="curation-explanation" id="alignments" *ngIf="curated && curated.length > 0">
  <h3>Curated alignments</h3>
  <p>
    Our virus sequencing datasets combine genomic sequences generated as part of our consortium with publicly available sequences. In addition to providing the raw data, we also curate sequences and provide the alignments to the
    broader community for downstream analyses.
  </p>

  <section id="download-alignments">
    <h5>Download the alignments</h5>
    <div class="download-data" *ngFor="let download of curated">
      <a [href]="download.contentUrl" target="_blank" rel="noreferrer">{{ download.description }}</a>
    </div>
  </section>

  <mat-expansion-panel id="methodology" [expanded]="panelOpenState" (opened)="panelOpenState = true" (closed)="panelOpenState = false">
    <mat-expansion-panel-header [collapsedHeight]="expansionPanelHeight" [expandedHeight]="expansionPanelHeight" class="methodology-header">
      <mat-panel-title class="methodology-title">
        <h5>{{ panelOpenState ? "Alignment methodology" : "View alignment methodology" }}</h5>
      </mat-panel-title>
    </mat-expansion-panel-header>
    <div class="methodology-description" *ngIf="includedInDataset == 'ebola-virus-seq'">
      <div>We periodically combine genomic sequences generated as part of our consortium with publicly available sequences. Our curated alignments do not include:
      </div>
      <ul>
        <li>
          laboratory strains (adapted, passaged, recombinant, antiviral & vaccine experiments)
        </li>
        <li>sequences without a timestamp</li>
        <li>subsequent timepoints, if multiple timepoints are available</li>
        <li>duplicates (when more than one sequence is available for a single strain)</li>

      </ul>
      <div>
        Remaining sequences are trimmed to their coding regions, codon aligned using MAFFT and inspected manually. At this step we discard:
      </div>
      <ul>
        <li>
          low quality sequences (manual curation)
        </li>
        <li>
          incomplete sequences (<95% of (NP+VP35+VP40+GP+VP30+VP24+L) ORFs length) </li> </ul> <div>
            <div>
              ORFs are arranged in sense orientation as follows:
            </div>
            <div>
              Ebola: NP -NNN- VP35 -NNN- VP40 -NNN- GP -NNN- VP30 -NNN- VP24 -NNN- L
            </div>
    </div>
</div>
<div class="methodology-description" *ngIf="includedInDataset == 'lassa-virus-seq'">
  <div>We periodically combine genomic sequences generated as part of our consortium with publicly available sequences. Our curated alignments do not include:</div>
  <ul>
    <li>
      laboratory strains (adapted, passaged, recombinant, antiviral & vaccine experiments)
    </li>
    <li>
      sequences without a timestamp
    </li>
    <li>
      subsequent timepoints, if multiple timepoints are available
    </li>
    <li>
      duplicates (when more than one sequence is available for a single strain)
    </li>
    <li>
      sequences from identified hospital epidemics
    </li>
  </ul>
  <div>
    Remaining sequences are trimmed to their coding regions, codon aligned using MAFFT and inspected manually. At this step we discard:
  </div>
  <ul>
    <li>
      low quality sequences (manual curation)
    </li>
    <li>
      incomplete sequences (<95% of (GP+NP) or (Z+L) ORFs length) </li> </ul> <div>
        ORFs are arranged in sense orientation as follows:
</div>
<div>
  Lassa S segment: NP -NNN- GP
</div>
<div>
  Lassa L segment: L -NNN- Z
</div>
</div>
</mat-expansion-panel>
</div>
